39 update for tns 2.0

jnation3406 opened the pull request:
Somehow forgot to make a PR for this when I changed it last week. This should fix tom_tns for TNS 2.0.

jchate6 reviewed (changes requested):
It looks like the Flux/Mag units are no longer valid.
A quick look at the forms on TNS (Broken links not withstanding) imply that the list of units is quite long (and general) and that AB Mag/ Vega Mag are now 1 word.

jnation3406 commented:
> It looks like the Flux/Mag units are no longer valid. A quick look at the forms on TNS (Broken links not withstanding) imply that the list of units is quite long (and general) and that AB Mag/ Vega Mag are now 1 word.

Thanks for catching that... I'll fix that here and in Hermes since I didn't catch that among there other values changes.

jnation3406 commented:
> > It looks like the Flux/Mag units are no longer valid. A quick look at the forms on TNS (Broken links not withstanding) imply that the list of units is quite long (and general) and that AB Mag/ Vega Mag are now 1 word.
> 
> Thanks for catching that... I'll fix that here and in Hermes since I didn't catch that among there other values changes.

Actually @jchate6 I looked through the code more and it looks correct. If you are setup to share through hermes, then it uses "AB mag" which maps in hermes to the correct index for TNS (1), if you have tom_tns setup to share through tns directly, then the options should be the TNS units, which should be a long list including "ABmag" and "VegaMag"

jchate6 commented:
I have `'ENABLE_TNS': False` in my hermes sharing.

When I made the previous comment, this submission didn't crash but resulted in an error saying that the mag units for the flux were not recognized.
Now I get the following:
![image](https://github.com/user-attachments/assets/ae583257-8149-4e12-82d4-145ecb5a81a7)
```
Internal Server Error: /tns/4/report
Traceback (most recent call last):
  File "/home/jchatelain/test_tom/demo_tom/env/lib64/python3.12/site-packages/django/core/handlers/exception.py", line 55, in inner
    response = get_response(request)
               ^^^^^^^^^^^^^^^^^^^^^
  File "/home/jchatelain/test_tom/demo_tom/env/lib64/python3.12/site-packages/django/core/handlers/base.py", line 199, in _get_response
    response = self.process_exception_by_middleware(e, request)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jchatelain/test_tom/demo_tom/env/lib64/python3.12/site-packages/django/core/handlers/base.py", line 362, in process_exception_by_middleware
    response = middleware_method(request, exception)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jchatelain/git/tom_base/tom_common/middleware.py", line 29, in process_exception
    raise exception
  File "/home/jchatelain/test_tom/demo_tom/env/lib64/python3.12/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jchatelain/test_tom/demo_tom/env/lib64/python3.12/site-packages/django/views/generic/base.py", line 104, in view
    return self.dispatch(request, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jchatelain/test_tom/demo_tom/env/lib64/python3.12/site-packages/django/views/generic/base.py", line 143, in dispatch
    return handler(request, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jchatelain/test_tom/demo_tom/env/lib64/python3.12/site-packages/django/views/generic/edit.py", line 153, in post
    return self.form_valid(form)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/jchatelain/git/tom_tns/tom_tns/views.py", line 87, in form_valid
    hermes_report, files = form.generate_hermes_report()
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jchatelain/git/tom_tns/tom_tns/forms.py", line 206, in generate_hermes_report
    'transient_type': dict(self.fields['at_type'].choices)[int(self.cleaned_data['at_type'])],
                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
KeyError: 1
Internal Server Error: /tns/4/report
Traceback (most recent call last):
  File "/home/jchatelain/test_tom/demo_tom/env/lib64/python3.12/site-packages/django/core/handlers/exception.py", line 55, in inner
    response = get_response(request)
               ^^^^^^^^^^^^^^^^^^^^^
  File "/home/jchatelain/test_tom/demo_tom/env/lib64/python3.12/site-packages/django/core/handlers/base.py", line 199, in _get_response
    response = self.process_exception_by_middleware(e, request)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jchatelain/test_tom/demo_tom/env/lib64/python3.12/site-packages/django/core/handlers/base.py", line 362, in process_exception_by_middleware
    response = middleware_method(request, exception)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jchatelain/git/tom_base/tom_common/middleware.py", line 29, in process_exception
    raise exception
  File "/home/jchatelain/test_tom/demo_tom/env/lib64/python3.12/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jchatelain/test_tom/demo_tom/env/lib64/python3.12/site-packages/django/views/generic/base.py", line 104, in view
    return self.dispatch(request, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jchatelain/test_tom/demo_tom/env/lib64/python3.12/site-packages/django/views/generic/base.py", line 143, in dispatch
    return handler(request, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jchatelain/test_tom/demo_tom/env/lib64/python3.12/site-packages/django/views/generic/edit.py", line 153, in post
    return self.form_valid(form)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/jchatelain/git/tom_tns/tom_tns/views.py", line 87, in form_valid
    hermes_report, files = form.generate_hermes_report()
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jchatelain/git/tom_tns/tom_tns/forms.py", line 206, in generate_hermes_report
    'transient_type': dict(self.fields['at_type'].choices)[int(self.cleaned_data['at_type'])],
                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
KeyError: 1

```

jnation3406 commented:
@jchate6 What happened to you should not have been possible - I'm guessing you have a cache in your test TOM, and since we cache the populated values, you were getting them cached from hermes and then turned submission through hermes off.

There seemed to be a bug in getting the values through TNS, which I just fixed by changing the tns_base_url to be just the base url and not `/api` on the end, and updating everywhere its used to use urljoin(). This is because urljoin('https://url.com/api', '/get/values') returns 'https://url.com/get/values' - notice the lack of the api in there. As long as you just use the base url + all appeneded parts with urljoin it is consistent and does what you expect, so that's why I've made this change.

On a side note - I don't really want to take on maintaining raw TNS support since I'd rather everyone just submit to the TNS through hermes, and spend my time making sure that pathway works.

jchate6 commented:
@jnation3406 I got the full list in a new tom. not sure how that happened.

jchate6 commented:
@jnation3406 just to keep you in the loop:

- In the `TNSReportForm` I've set `submitter` to required=False to match the classification form since I believe this is only required for hermes, and Hermes will give it's on validation.
- I'm currently struggling with `"reporting_groupid"` and how we translate from a given group to a group ID. I'll work on figuring that part out tomorrow.

Thanks for helping with this, I'll try to get the direct TNS side working and ask for a review from you if that's ok?

jnation3406 commented:
@jchate6 Sure - reporting group id comes from the tns values api in the 'groups' section. There is a mapping in the tns_api.py file. Note that in TNS 2.0 changes were made so you now can only report to groups your bot is a member of. Previously you could report to any group - so if you get weird errors on submission make sure your bot is a member of the group you are trying to report to. The error you will get does not give you that hint.

jchate6 commented:
@jnation3406 do you have any idea why hermes_bot is not in Hermes_group??
What else do I need to do?
![image](https://github.com/user-attachments/assets/2da48f9c-c9f6-44d8-8556-47dbcd2c5989)
![image](https://github.com/user-attachments/assets/1dcbfcbe-41e2-446b-a2bd-bea8934edf59)
![image](https://github.com/user-attachments/assets/169e340f-51e7-4138-b46a-9b7dbee47982)

jnation3406 commented:
@jchate6 Sorry about that - it is a TNS 2.0 change. I had added it in the sandbox but apparently not the real site. I've added it back to both now so it should be in the group.
